### PR TITLE
Treeowl map maybe m

### LINKED ISSUE #183: mapMaybeM
Can I add `mapMaybeM` to `vector`? The type signature would be:

    mapMaybeM :: (a -> m (Maybe b)) -> Vector a -> m (Vector b)

Although the list-equivalent of `mapMaybeM` is not in `base`, it's a pretty common thing for people to write. I can see from some googling that it exists in [`monad-extras`](https://hackage.haskell.org/package/monad-extras-0.6.0/docs/Control-Monad-Extra.html#v:mapMaybeM), [`extra`](https://hackage.haskell.org/package/extra-1.6/docs/Control-Monad-Extra.html#v:mapMaybeM), and in [`stack`](https://hackage.haskell.org/package/stack-1.0.0/docs/src/Data.Maybe.Extra.html). I bring this up because I think it highlights the utility of such a function and the common expected semantics people seem to have for it.

### PR DESCRIPTION
**Pull request by Shimuuar**
This is #226 rebased on top of master and with haddocks. Only think holding #226 is lack of documentation so I think now it's ready to merge. 

Fixes #183

### CODE REVIEW

File: Data/Vector/Generic.hs [hunk 1]
```diff
@@ -1372,6 +1372,18 @@ filterM :: (Monad m, Vector v a) => (a -> m Bool) -> v a -> m (v a)
...
+-- discard elements returning Nothing.
+mapMaybeM :: (Monad m, Vector v a, Vector v b) => (a -> m (Maybe b)) -> v a -> m (v b)
+{-# INLINE mapMaybeM #-}
+mapMaybeM f = unstreamM . Bundle.mapMaybeM f . stream
+
+-- | /O(n)/ Apply monadic function to each element of vector and its index.
+-- Discards elements returning Nothing.
+imapMaybeM :: (Monad m, Vector v a, Vector v b)
+      => (Int -> a -> m (Maybe b)) -> v a -> m (v b)
+imapMaybeM f = unstreamM . Bundle.mapMaybeM (\(i, a) -> f i a) . Bundle.indexed . stream
```
Comment (lehins) [inline Data/Vector/Generic.hs:1385]: Missing the inline pragma:
```suggestion
{-# INLINE imapMaybeM #-}
imapMaybeM f = unstreamM . Bundle.mapMaybeM (\(i, a) -> f i a) . Bundle.indexed . stream
```

File: Data/Vector/Generic.hs [hunk 2]
```diff
@@ -99,7 +99,7 @@ module Data.Vector.Generic (
   -- ** Filtering
   filter, ifilter, uniq,
   mapMaybe, imapMaybe,
-  filterM,
+  filterM, mapMaybeM, imapMaybeM,
```
Comment (lehins) [inline Data/Vector/Generic.hs:102]: Not terribly important, but for consistency sake. All other modules export `mapMaybeM` and `imapMaybeM` before `filterM` and after `mapMaybeM`, `imapMaybe`.

```suggestion
  mapMaybeM, imapMaybeM,
  filterM,
```

**Review by lehins — APPROVED** [review]
Couple of minor comments, but otherwise looks great.

File: Data/Vector/Generic.hs (hunk 2 above)
Comment (Shimuuar) [reply Data/Vector/Generic.hs]: Data.Vector wasn't consistent as well. I arranged everything in the same way. I also put `filterM` after `filter` & `ifilter`. I think it makes sense to keep all filter variants together.

**Review by Bodigrim — APPROVED** [review]

**Pull request merged**